lib/AuthStateManager.ts

```diff
@@ -0,0 +1,122 @@
+import { OktaAuth } from './browser';
+import { ACCESS_TOKEN_STORAGE_KEY, ID_TOKEN_STORAGE_KEY } from './constants';
+const PCancelable = require('p-cancelable');
+
+const DEFAULT_AUTH_STATE = { 
+  isPending: true,
+  isAuthenticated: false,
+  idToken: null,
+  accessToken: null,
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+
+class AuthStateManager {
+  _sdk: OktaAuth;
+  _pending: { updateAuthStatePromise: typeof PCancelable };
+  _authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this._sdk = sdk;
+    this._pending = { updateAuthStatePromise: null };
+    this._authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to sync update tokens in memory (this.pending), and start updateState process
+    // "added" event is emitted in both add and renew process
+    // Only listen on "added" (instead of both "added" and "renewed") to limit authState re-evaluation
+    sdk.tokenManager.on('added', (key, token) => {
+      this.updateAuthState({ shouldCheckExpiration: false,  event: 'added', key, token });
+    });
+    sdk.tokenManager.on('renewed', (key, token) => {
+      this.updateAuthState({ shouldCheckExpiration: false,  event: 'renewed', key, token });
+    });
+    sdk.tokenManager.on('removed', (key, token) => {
+      this.updateAuthState({ shouldCheckExpiration: false, event: 'removed', key, token });
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this._authState;
+  }
+
+  updateAuthState({ shouldCheckExpiration, event, key, token }: UpdateAuthStateOptions): void {
+    const logger = (status) => {
+      console.group(`OKTA-AUTH-JS:updateAuthState:event:${event}:${status}`);
+      console.log(key, token);
+      console.log('Current authState', this._authState);
+      console.groupEnd();
+    };
+
+    const emitAuthStateChange = (authState) => {
+      this._authState = authState;
+      // emit new authState object
+      this._sdk.emitter.emit(EVENT_AUTH_STATE_CHANGE, { ...authState });
+      if (devMode) {
+        logger('sent');
+      }
+    };
+
+    if (this._pending.updateAuthStatePromise) {
+      this._pending.updateAuthStatePromise.cancel();
+    }
+
+    const { isAuthenticated, devMode } = this._sdk.options;
+    const cancelablePromise = new PCancelable((resolve, _, onCancel) => {
+      onCancel.shouldReject = false;
+      onCancel(() => {
+        this._pending.updateAuthStatePromise = null;
+        if (devMode) {
+          logger('canceled');
+        }
+      });
+
+      return Promise.all([
+        this._sdk.tokenManager.get(ACCESS_TOKEN_STORAGE_KEY),
+        this._sdk.tokenManager.get(ID_TOKEN_STORAGE_KEY)
+      ]).then(([accessToken, idToken]) => {
+        if (cancelablePromise.isCanceled) {
+          resolve();
+          return;
+        }
+
+        if (shouldCheckExpiration) {
```

Comment:
Looks like we still need to keep a flag like this to distinguish between initial update and event-driven updates. As when a renew process start, it add a new token into storage, then start to evaluate the authState. It can run into the scenario where the unrenewed token in storage is also logically expired, then it will be used as `null` in the evaluation, which provides the false-positive information to the downstream. Then the state will be corrected by the followed renew process, but it will cause multiple renders in the app, since the isAuthenticated state changed from `true -> false -> true`.

So then public APIs from OktaAuth can be both `initialAuthState` and `updateAuthState` as we don't want the users to touch args directly.
https://github.com/okta/okta-auth-js/pull/453/files#diff-75082c639aefcc47b3b8df57c8870867R333

Reply:
> Looks like we still need to keep a flag like this to distinguish between initial update and event-driven updates

Why a flag?  Why not simply kick off an update when the code is first run (object initialization)?

Reply:
You mean in the AuthJs' constructor? I also thought that way, but it will cause a problem for the event listener in the downstream SDKs to miss the first initial event to update the UI.

Reply:
My initial thought is that, within `updateAuthState` we want to not only check if the tokens are expired, but also call `tokenManager.renew` if they are expired and `options.autoRenew` is true.  This will defer the emitted authState until after token renew is complete. There may be a better way to accomplish this, but logically we do want to wait for the initial autoRenew cycle to complete before we emit initial authState. Maybe we can handle it in our initialization of autoRenew logic. If autoRenew is false, do initial authState immediately. If autoRenew is true, do autoRenew **then** do initial authState

Reply:
The UI can grab getAuthState on first render, and if anything is pending it will get the events later (and if it's not pending, great, it has the state)

This is different from okta-react today, which is triggering an updateAuthState on first render.

Reply:
Initial false-positive is handled here: https://github.com/okta/okta-auth-js/pull/453/files#diff-f4d380d895ef82fba53a76eeed9089e1R84
As we are using storage as the source of truth, the change clear expired token in storage synchronously without triggering an event, then trigger expired event to start the renew/remove flow.
Also, removed token dependency for the renew process.

With this change, TokenManager should be able to just keep the valid tokens in storage (in the flows that under the SDK's control), then AuthStateManager can trust on tokens from storage w/o expiration check.